components/script/dom/htmltextareaelement.rs

```diff
@@ -232,10 +232,25 @@ impl HTMLTextAreaElementMethods for HTMLTextAreaElement {
 
     // https://html.spec.whatwg.org/multipage/#dom-textarea-value
     fn SetValue(&self, value: DOMString) {
-        // TODO move the cursor to the end of the field
-        self.textinput.borrow_mut().set_content(value);
+        let mut textinput = self.textinput.borrow_mut();
+
+        // Step 1
+        let old_value = textinput.get_content();
+        let old_selection = textinput.selection_begin;
+
+        // Step 2
+        textinput.set_content(value);
+
+        // Step 3
         self.value_changed.set(true);
 
+        if old_value != textinput.get_content() {
+            // Step 4
+            textinput.adjust_horizontal_to_limit(Direction::Forward, Selection::NotSelected);
+        } else {
+            textinput.selection_begin = old_selection;
```

Comment:
This is necessary because `textinput.set_content` resets the `selection_begin` field. But maybe it would be better to change `set_content` so that it doesn't do that unless necessary? (I.e. unless `selection_begin` now points to an invalid location.) It already does something similar with `edit_point`.

Reply:
By the way, we could just not call `set_content` at all if `old_value == value`. But I didn't go for that because the spec differentiates the "raw value" and the "API value". This differentiation isn't made in the code, yet, but I presumed that in the future we may want to call `set_content`, and then pull out the "API value" for the comparison. Maybe this is a premature assumption though.

Reply:
This looks fine to me. If the API value has not changed, `selection_begin` shouldn't point to any invalid location at all.

Reply:
So, I think for our implementation, we don't make the distinction between API values and raw values, because in `set_content`, we split the input DOMString via LF into multiple Strings in a Vec, and when we call `get_content`, the function knows to append LF to every String in the Vec, and return the result as a single DOMString.

However, [looking at the algorithm to get an API value from `HTMLTextAreaElement`](https://html.spec.whatwg.org/multipage/form-elements.html#the-textarea-element:concept-fe-api-value-2), I can see that we currently don't split CRLF or CR in `set_content`. So perhaps we should change that and ensure it does split on CRLF and CR.

Reply:
> This looks fine to me. If the API value has not changed, selection_begin shouldn't point to any invalid location at all.

Yep, the implementation works fine, I just felt like changing `set_content` would make the code here a bit cleaner. I don't really mind either way though.

> So perhaps we should change that and ensure it does split on CRLF and CR.

Yes, definitely extra handling for this will be needed in order to follow the spec. I haven't tried to address that in this PR but I agree it will need to be done. There are already some tests for this stuff in `tests/wpt/web-platform-tests/html/semantics/forms/the-textarea-element/value-defaultValue-textContent.html`. I think that it may not be enough to make `set_content` also split on CR/CRLF, as that would lose the original "raw value".

Reply:
I don't think that's an important concern, since every instance of the raw value is being converted into an API value, or a value, or that it's being used for comparison, in which case you can also use the corresponding API value or value to compare.

Reply:
This needs to be addressed.

Reply:
Ok, I felt like that was a separate change really, but I can take a look